provider/image.go

```diff
@@ -642,3 +654,118 @@ func processLogLine(msg string) (string, error) {
 	}
 	return info, nil
 }
+
+func configureDockerClient(configs map[string]string) (*client.Client, error) {
+	// check for TLS inputs
+	var caMaterial, certMaterial, keyMaterial, certPath, host string
+	if val, ok := configs["caMaterial"]; ok {
+		caMaterial = val
+	}
+	if val, ok := configs["certMaterial"]; ok {
+		certMaterial = val
+	}
+	if val, ok := configs["keyMaterial"]; ok {
+		keyMaterial = val
+	}
+	if val, ok := configs["certPath"]; ok {
+		certPath = val
+	}
+	if val, ok := configs["host"]; ok {
+		host = val
+	}
+
+	if certMaterial != "" || keyMaterial != "" {
+		if certMaterial == "" || keyMaterial == "" {
+			return nil, fmt.Errorf("certMaterial, and keyMaterial must be specified")
+		}
+
+		if certPath != "" {
+			return nil, fmt.Errorf("when using raw certificates, certPath must not be specified")
+		}
+		httpClient, err := buildHTTPClientFromBytes([]byte(caMaterial), []byte(certMaterial), []byte(keyMaterial))
+		if err != nil {
+			return nil, err
+		}
+
+		// Set custom client first
+		return client.NewClientWithOpts(
+			client.WithHTTPClient(httpClient),
+			client.FromEnv,
+			client.WithHost(host),
+			client.WithAPIVersionNegotiation(),
+		)
+	}
+
+	var ca, cert, key string
+	if certPath != "" {
+		ca = filepath.Join(certPath, "ca.pem")
+		cert = filepath.Join(certPath, "cert.pem")
+		key = filepath.Join(certPath, "key.pem")
+		return client.NewClientWithOpts(
+			client.FromEnv,
+			client.WithTLSClientConfig(ca, cert, key),
+			client.WithAPIVersionNegotiation(),
+		)
+	}
+
+	return client.NewClientWithOpts(
+		client.FromEnv,
+		client.WithAPIVersionNegotiation(),
+	)
+}
+
+// buildHTTPClientFromBytes builds the http client from bytes (content of the files)
+func buildHTTPClientFromBytes(caPEMCert, certPEMBlock, keyPEMBlock []byte) (*http.Client, error) {
+	tlsConfig := &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		MaxVersion: 0,
+	}
+	if certPEMBlock != nil && keyPEMBlock != nil {
+		tlsCert, err := tls.X509KeyPair(certPEMBlock, keyPEMBlock)
+		if err != nil {
+			return nil, err
+		}
+		tlsConfig.Certificates = []tls.Certificate{tlsCert}
+	}
+
+	if len(caPEMCert) == 0 {
+		tlsConfig.InsecureSkipVerify = true
```

Comment:
I don't think we should skip TLS verification when the CA is missing and instead we should fail since it has security implications.

That said, maybe we should create a feature request to implement a boolean on TLS Verification to independently set this flag. https://www.mankier.com/1/docker#--tlsverify

Reply:
I see we have a preference for being stricter here. Let's do that for now and if there's requests to allow skipping, we can add them back in. 👍